File: sample_app/src/uk/co/alt236/btlescan/util/BluetoothLeDeviceStore.java
package uk.co.alt236.btlescan.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import uk.co.alt236.bluetoothlelib.device.BluetoothLeDevice;

public class BluetoothLeDeviceStore {
	private final Map<String, BluetoothLeDevice> mDeviceMap;

	public BluetoothLeDeviceStore(){
		mDeviceMap = new HashMap<String, BluetoothLeDevice>();
	}

	public void addDevice(BluetoothLeDevice device){
		if(mDeviceMap.containsKey(device.getAddress())){

		} else {
			mDeviceMap.put(device.getAddress(), device);
		}
	}


	public List<BluetoothLeDevice> getDeviceList(){
		final List<BluetoothLeDevice> methodResult = new ArrayList<BluetoothLeDevice>(mDeviceMap.values());


		Collections.sort(methodResult, new Comparator<BluetoothLeDevice>() {

			@Override
			public int compare(BluetoothLeDevice arg0, BluetoothLeDevice arg1) {
				return arg0.getAddress().compareToIgnoreCase(arg1.getAddress());
			}
		});


		return methodResult;
	}
}

File: sample_app/src/main/java/uk/co/alt236/btlescan/bluetooth/BluetoothLeDeviceStore.kt
package uk.co.alt236.btlescan.bluetooth

import dev.alt236.bluetoothlelib.device.BluetoothLeDevice
import dev.alt236.easycursor.objectcursor.EasyObjectCursor
import java.util.Collections

class BluetoothLeDeviceStore {
    private val mDeviceMap = HashMap<String, BluetoothLeDevice>()

    fun addDevice(device: BluetoothLeDevice) {
        if (mDeviceMap.containsKey(device.address)) {
            mDeviceMap[device.address]!!.updateRssiReading(device.timestamp, device.rssi)
        } else {
            mDeviceMap[device.address] = device
        }
    }

    fun clear() {
        mDeviceMap.clear()
    }

    val size: Int
        get() = mDeviceMap.size

    val deviceCursor: EasyObjectCursor<BluetoothLeDevice>
        get() = getDeviceCursor(DEFAULT_COMPARATOR)

    fun getDeviceCursor(comparator: Comparator<BluetoothLeDevice>): EasyObjectCursor<BluetoothLeDevice> =
        EasyObjectCursor(
            BluetoothLeDevice::class.java,
            getDeviceList(comparator),
            "address",
        )

    val deviceList: List<BluetoothLeDevice>
        get() = getDeviceList(DEFAULT_COMPARATOR)

    fun getDeviceList(comparator: Comparator<BluetoothLeDevice>): List<BluetoothLeDevice> {
        val methodResult: List<BluetoothLeDevice> = ArrayList(mDeviceMap.values)
        Collections.sort(methodResult, comparator)
        return methodResult
    }

    private class BluetoothLeDeviceComparator : Comparator<BluetoothLeDevice> {
        override fun compare(
            arg0: BluetoothLeDevice,
            arg1: BluetoothLeDevice,
        ): Int = arg0.address.compareTo(arg1.address)
    }

    companion object {
        private val DEFAULT_COMPARATOR = BluetoothLeDeviceComparator()
    }
}
